### ISSUE TITLE
Model save repetition causes threading synchronization issue

### ISSUE THREAD

**Issue by jbed94** [tensorflow/tensorflow#56775]
Hi everyone,

## Short description
I have encountered a problem with saving a model, or rather overriding a saved model. It seems that the problem is somehow connected with threading. Below a paste a source code to reproduce, and logs:

## Source code
```
import tensorflow as tf

path = './checkpoint/model'

model = tf.keras.Sequential([
    tf.keras.layers.Dense(100),
    tf.keras.layers.Dense(100)
])

model(tf.zeros([1, 100]))

checkpoint = tf.train.Checkpoint(model=model)
manager = tf.train.CheckpointManager(checkpoint, directory=path, max_to_keep=5)

manager.save()
print('saved1')
manager.save()  # here an error occurs
print('saved2')
```

## Errors:
The errors I get are (every script execution causes randomly a different error, thus I expect that it's synch issue):
```
terminate called after throwing an instance of 'std::system_error'
  what():  Invalid argument
```
or
```
terminate called after throwing an instance of 'std::system_error'
  what():  Owner **died**
```
or simply without error but with exit code:
```
Process finished with exit code 136 (interrupted by signal 8: SIGFPE)
```
or:
```
python: ../nptl/pthread_mutex_lock.c:428: __pthread_mutex_lock_full: Assertion `e != ESRCH || !robust' failed.
```
or
```
Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)
```
or:
```
The futex facility returned an unexpected error code.
```

## System and libraries

The TensorFlow package that I use is compiled by myself in r2.9 version for GPU (RTX 3060 12GB, 8.6 CC, cudnn 8.4, cuda 11.6, GCC 9.3). The CPU is: AMD® Ryzen 9 5950x 16-core processor × 32

TensorFlow compile command:
```
bazel build --jobs=16 --config=nogcp --config=numa --config=monolithic --config=cuda //tensorflow/tools/pip_package:build_pip_package
```

## Help
Does anyone have any idea what the problem may be? When I compiled tensorflow in a similar way on another computer (laptop) everything works fine.

**Comment by jbed94** [comment]
Hello again, after a couple of TensorFlow recompilations I finally managed to solve the issue (or at least I found a build config that works fine). It seems that --config=monilithic was the source of the problem. After removing that one TF works fine. Below I paste the command for successful compilation:

```
bazel build --jobs=16 --config=nogcp -c opt --copt=-mavx --copt=-mavx2 --copt=-mfma --copt=-mfpmath=both --copt=-msse4.2 --config=cuda --config=numa //tensorflow/tools/pip_package:build_pip_package
```

Anyway, it still seems worthful of finding out why a monolithic build (which afaik should disable the possibility for building custom operations only) causes such a strange threading error.
